addons/product/models/product_attribute.py

```diff
@@ -128,19 +181,47 @@ class ProductTemplateAttributeLine(models.Model):
     _description = 'Product Template Attribute Line'
     _order = 'attribute_id, id'
 
-    product_tmpl_id = fields.Many2one('product.template', string='Product Template', ondelete='cascade', required=True, index=True)
-    attribute_id = fields.Many2one('product.attribute', string='Attribute', ondelete='restrict', required=True, index=True)
-    value_ids = fields.Many2many('product.attribute.value', string='Attribute Values')
+    product_tmpl_id = fields.Many2one('product.template', string="Product Template", ondelete='cascade', required=True, index=True,
+        states={'created': [('readonly', True)]})
+    attribute_id = fields.Many2one('product.attribute', string="Attribute", ondelete='restrict', required=True, index=True,
+        states={'created': [('readonly', True)]})
+    value_ids = fields.Many2many('product.attribute.value', string="Values", domain="[('attribute_id', '=', attribute_id)]",
+        relation='product_attribute_value_product_template_attribute_line_rel')
     product_template_value_ids = fields.Many2many(
         'product.template.attribute.value',
         string='Product Attribute Values',
         compute="_set_product_template_value_ids",
         store=False)
+    state = fields.Selection(compute='_compute_state',
+        selection=[
+            ('draft', "Draft"),
+            ('created', "Created"),
+        ],
+        default='draft',
+    )
+
+    def _compute_state(self):
+        for ptal in self:
+            ptal.state = 'created' if isinstance(ptal.id, int) else 'draft'
+
+    @api.onchange('attribute_id')
+    def _onchange_attribute_id(self):
+        self.value_ids = self.value_ids.filtered(lambda pav: pav.attribute_id == self.attribute_id)
 
     @api.constrains('value_ids', 'attribute_id')
-    def _check_valid_attribute(self):
-        if any(not line.value_ids or line.value_ids > line.attribute_id.value_ids for line in self):
-            raise ValidationError(_('You cannot use this attribute with the following value.'))
+    def _check_valid_values(self):
+        for ptal in self:
+            if not ptal.value_ids:
+                raise ValidationError(
+                    _("The attribute <strong>%s</strong> must have at least one value for the product %s.") %
+                    (ptal.attribute_id.display_name, ptal.product_tmpl_id.display_name)
+                )
+            for pav in ptal.value_ids:
```

Comment:
```python
wrong_attr = ptal.value_ids - ptal.attribute_id.value_ids
if wrong_attr:
  raise ValidationError(_('On the product %s, you cannot use the attribute %s with the values %s.') % (ptal.product_tmpl_id.display_name, ptal.attribute_id.name, wrong_attr.mapped('name')))
```
Wouldn't it be clearer as only one error will be raise even if multiple values are wrong ?

Reply:
As discussed we will keep it as it is now, because as a general principle we prefer to fail fast when an exception occurs, and we don't want an error that is potentially too long.